_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,231 @@
+
+Let's start building.  Start a TypeScript project and create an `addressBook.proto` file as above in the same project. Then run:
+
+- `npm install ts-protoc-gen`
+
+Next compile your `.proto` file into a `.ts` file for use within your project by running the following CLI command:
+
+```
+protoc \
+    --plugin="./node_modules/.bin/protoc-gen-ts_proto" \
+    --ts_proto_opt=esModuleInterop=true \
+    --ts_proto_out="./src/generated" \
+    src/proto/addressBook.proto
+```
+
+This is the point where I really like TypeScript. If you run `protoc` without the `ts-proto` plugin and look at the generated code there are 760 lines of stuff in the `generated/src/proto/addressBook.js` file which honestly is not that easy to read or understand how to use. But if you look at the `generated/src/proto/addressBook.ts` file, there are only 250 lines and it provides you with the type interfaces and commands you are going to need to create and use your messages and Protobufs.
+
+_*This is where I fell over on my first attempt, I did not use a plugin like `ts-proto` to generate the types and only had the JS output. I couldn't easily enough understand what was going on, got lost and then distracted by other shiny stuff.*_
+
+## Create a Message
+_*When you read the docs and get presented with ["The API is not well-documented yet"](https://github.com/protocolbuffers/protobuf/tree/master/js#api) you can be forgiven for wanting to cry and giving up. This is where I failed on my second attempt and this is where the Typescript types come to the rescue.*_
+
+Creating the contents of your message is now as easy as:
+```js
+import { AddressBook, Person_PhoneType } from './generated/src/proto/addressBook'
+
+const myAddressBook: AddressBook = {
+  people: [
+    {
+      name: "Joe Blogs",
+      phones: [
+        {
+          phoneNumber: "0123456789",
+          phoneType: Person_PhoneType.MOBILE,
+        }
+      ]
+    },
+    {
+      name: "Jane Smith",
+      phones: [
+        {
+          phoneNumber: "0987654321",
+          phoneType: Person_PhoneType.HOME,
+        }
+      ]
+    }
+  ]
+}
+```
+This is almost identical to the original JSON object used in the [data structure section above](#define-the-data-structure-for-your-messages), how fantastic is that!  There are two differences in the above that take advantage of some awesome idiomatic Typescript and make life really easy:
+- The object type assignment `AddressBook` which allows you to use type checking for creating valid messages
+- The `phoneType` is assigned through an enum, e.g. `Person_PhoneType.MOBILE`
+
+However, if at this point you are thinking _*"Now wait up, this is not like the other tutorials I have seen. I was expecting to use a class object and function calls to create the message"*_, you can do it that way with Typescript, have a look at the [How to do this so it looks like most published tutorials](#how-to-do-this-so-it-looks-like-most-published-tutorials) section below.
```

Comment:
```suggestion
If at this point you are thinking _*"Now wait up, this is not like the other tutorials I have seen. I was expecting to use a class object and function calls to create the message"*_, you can do it that way with Typescript, have a look at the [How to do this so it looks like most published tutorials](#how-to-do-this-so-it-looks-like-most-published-tutorials) section below.
```
nitpick: https://blog.wordvice.com/avoid-fillers-powerful-writing/

There's probably more you could remove to make it more punchy - I'll leave this for you as a proof read if you want to.

Reply:
I like your use of style here - you have separate 'voices' in the prose that are contributing to the overall flow.

It flows well and also signposts content for the reader, making me pause for thought and reflect.

:+1:

Reply:
Kaisen